src/ring/middleware/session_timeout.clj

```diff
@@ -18,22 +31,33 @@
   :timeout-response - the response to send if an idle timeout occurs
   :timeout-handler  - the handler to run if an idle timeout occurs"
   {:arglists '([handler options])}
-  [handler {:keys [timeout timeout-response timeout-handler] :or {timeout 600}}]
+  [handler {:keys [timeout timeout-response timeout-handler] :or {timeout 600} :as options}]
   {:pre [(integer? timeout)
          (if (map? timeout-response)
            (nil? timeout-handler)
            (ifn? timeout-handler))]}
-  (fn [request]
-    (let [session  (:session request {})
-          end-time (::idle-timeout session)]
-      (if (and end-time (< end-time (current-time)))
-        (assoc (or timeout-response (timeout-handler request)) :session nil)
-        (when-let [response (handler request)]
-          (let [session (:session response session)]
-            (if (nil? session)
-              response
-              (let [end-time (+ (current-time) timeout)]
-                (assoc response :session (assoc session ::idle-timeout end-time))))))))))
+  (fn
+    ([request]
+     (idle-session-timeout request (handler request) options))
```

Comment:
The `handler` function should not be run if the timeout triggers.

Reply:
I broke out a function for checking if the session has expired and another for returning the expiration response with the goal of re-using these in the asynchronous variant. I believe I have both invoking the handler _only_ if the timeout has not expired.